File: trade-backend-go/internal/utils/path_manager.go
package utils

import (
	"log/slog"
	"os"
	"path/filepath"
)

// PathManager manages paths for configuration and data files.
// Exact conversion of Python PathManager class.
type PathManager struct {
	dataDir   string
	configDir string
	cacheDir  string
}

// NewPathManager creates a new path manager.
// Exact conversion of Python PathManager.__init__ method.
func NewPathManager() *PathManager {
	pm := &PathManager{}
	pm.dataDir = pm.determineDataDirectory()
	pm.configDir = filepath.Join(pm.dataDir, "config")
	pm.cacheDir = filepath.Join(pm.dataDir, "cache")
	
	pm.ensureDirectoryStructure()
	return pm
}

// determineDataDirectory determines the appropriate data directory based on environment.
// Exact conversion of Python _determine_data_directory method.
func (pm *PathManager) determineDataDirectory() string {
	containerDataDir := "/app/data"
	
	if _, err := os.Stat(containerDataDir); err == nil {
		slog.Info("📁 Using container data directory: " + containerDataDir)
		return containerDataDir
	} else {
		// Development mode - use parent directory (project root)
		// Go server runs from trade-backend-go, but config files are in parent directory
		currentDir, _ := os.Getwd()
		parentDir := filepath.Join(currentDir, "..")
		absParentDir, _ := filepath.Abs(parentDir)
		slog.Info("📁 Using development data directory: " + absParentDir)
		return absParentDir
	}
}

// ensureDirectoryStructure ensures required directory structure exists.
// Exact conversion of Python _ensure_directory_structure method.
func (pm *PathManager) ensureDirectoryStructure() {
	if pm.IsContainerMode() {
		// Only create subdirectories in container mode
		os.MkdirAll(pm.configDir, 0755)
		os.MkdirAll(pm.cacheDir, 0755)
		slog.Info("📁 Created directory structure in " + pm.dataDir)
	}
}

// DataDir gets the data directory path.
func (pm *PathManager) DataDir() string {
	return pm.dataDir
}

// ConfigDir gets the config directory path.
func (pm *PathManager) ConfigDir() string {
	return pm.configDir
}

// CacheDir gets the cache directory path.
func (pm *PathManager) CacheDir() string {
	return pm.cacheDir
}

// IsContainerMode checks if running in container mode.
// Exact conversion of Python is_container_mode method.
func (pm *PathManager) IsContainerMode() bool {
	return pm.dataDir == "/app/data"
}

// GetConfigFilePath gets the full path for a configuration file.
// Exact conversion of Python get_config_file_path method.
func (pm *PathManager) GetConfigFilePath(filename string) string {
	if pm.IsContainerMode() {
		return filepath.Join(pm.configDir, filename)
	} else {
		// Development mode - use root directory
		return filepath.Join(pm.dataDir, filename)
	}
}

// GetCacheFilePath gets the full path for a cache file.
// Exact conversion of Python get_cache_file_path method.
func (pm *PathManager) GetCacheFilePath(filename string) string {
	if pm.IsContainerMode() {
		return filepath.Join(pm.cacheDir, filename)
	} else {
		// Development mode - use cache subdirectory
		cacheDir := filepath.Join(pm.dataDir, "cache")
		os.MkdirAll(cacheDir, 0755)
		return filepath.Join(cacheDir, filename)
	}
}

// GetStatus gets current path configuration status.
// Exact conversion of Python get_status method.
func (pm *PathManager) GetStatus() map[string]interface{} {
	mode := "development"
	if pm.IsContainerMode() {
		mode = "container"
	}
	
	return map[string]interface{}{
		"mode":               mode,
		"data_dir":           pm.dataDir,
		"config_dir":         pm.configDir,
		"cache_dir":          pm.cacheDir,
		"data_dir_exists":    dirExists(pm.dataDir),
		"config_dir_exists":  dirExists(pm.configDir),
		"cache_dir_exists":   dirExists(pm.cacheDir),
	}
}

// Helper function to check if directory exists
func dirExists(path string) bool {
	if _, err := os.Stat(path); err == nil {
		return true
	}
	return false
}

// Global instance - exact same as Python
var GlobalPathManager = NewPathManager()

File: strategy-service/src/path_manager.py
from pathlib import Path
import logging

logger = logging.getLogger(__name__)

class PathManager:
    """
    Centralized path management for configuration and data files.
    
    Follows container best practices:
    - Container mode: Uses /app/data (when directory exists)
    - Development mode: Uses current directory (fallback)
    """
    
    def __init__(self):
        self._data_dir = self._determine_data_directory()
        self._config_dir = self._data_dir / "config"
        self._cache_dir = self._data_dir / "cache"
        
        # Ensure directory structure exists
        self._ensure_directory_structure()
    
    def _determine_data_directory(self) -> Path:
        """Determine the appropriate data directory based on environment."""
        container_data_dir = Path("/app/data")
        
        if container_data_dir.exists():
            logger.info(f"📁 Using container data directory: {container_data_dir}")
            return container_data_dir
        else:
            # Development mode - use project root (parent of strategy-service)
            # This ensures parquet data at /project/parquet is accessible
            project_root = Path(__file__).parent.parent.parent
            logger.info(f"📁 Using development data directory: {project_root}")
            return project_root
    
    def _ensure_directory_structure(self):
        """Ensure required directory structure exists."""
        if self.is_container_mode():
            # Only create subdirectories in container mode
            self._config_dir.mkdir(parents=True, exist_ok=True)
            self._cache_dir.mkdir(parents=True, exist_ok=True)
            logger.info(f"📁 Created directory structure in {self._data_dir}")
    
    @property
    def data_dir(self) -> Path:
        """Get the data directory path."""
        return self._data_dir
    
    @property
    def config_dir(self) -> Path:
        """Get the config directory path."""
        return self._config_dir
    
    @property
    def cache_dir(self) -> Path:
        """Get the cache directory path."""
        return self._cache_dir
    
    def is_container_mode(self) -> bool:
        """Check if running in container mode."""
        return self._data_dir == Path("/app/data")
    
    def get_config_file_path(self, filename: str) -> Path:
        """Get the full path for a configuration file."""
        if self.is_container_mode():
            return self._config_dir / filename
        else:
            # Development mode - use root directory
            return self._data_dir / filename
    
    def get_cache_file_path(self, filename: str) -> Path:
        """Get the full path for a cache file."""
        if self.is_container_mode():
            return self._cache_dir / filename
        else:
            # Development mode - use cache subdirectory
            cache_dir = self._data_dir / "cache"
            cache_dir.mkdir(exist_ok=True)
            return cache_dir / filename
    
    def get_status(self) -> dict:
        """Get current path configuration status."""
        return {
            "mode": "container" if self.is_container_mode() else "development",
            "data_dir": str(self._data_dir),
            "config_dir": str(self._config_dir),
            "cache_dir": str(self._cache_dir),
            "data_dir_exists": self._data_dir.exists(),
            "config_dir_exists": self._config_dir.exists(),
            "cache_dir_exists": self._cache_dir.exists()
        }

# Global instance
path_manager = PathManager()
